daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/UpgradesIT.scala

```diff
@@ -0,0 +1,365 @@
+// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.engine.script
+package test
+
+import io.circe._
+import io.circe.yaml
+import java.io.File
+import java.nio.file.{Files, Path, Paths}
+import com.daml.bazeltools.BazelRunfiles.{requiredResource, rlocation}
+import com.daml.lf.data.Ref._
+import com.daml.lf.engine.script.ScriptTimeMode
+import com.daml.lf.engine.script.test.DarUtil.{buildDar, Dar, DataDep}
+import com.daml.lf.language.{LanguageMajorVersion, LanguageVersion}
+import com.daml.lf.engine.script.v2.ledgerinteraction.grpcLedgerClient.test.TestingAdminLedgerClient
+import com.daml.scalautil.Statement.discard
+import com.daml.timer.RetryStrategy
+import com.digitalasset.canton.ledger.client.configuration.LedgerClientChannelConfiguration
+import org.scalatest.Inside
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AsyncWordSpec
+import scala.concurrent.Future
+import scala.sys.process._
+import scala.util.matching.Regex
+import scala.collection.mutable
+import scala.concurrent.duration.DurationInt
+
+class UpgradesIT extends AsyncWordSpec with AbstractScriptTest with Inside with Matchers {
+
+  final override protected lazy val nParticipants = 2
+  final override protected lazy val timeMode = ScriptTimeMode.WallClock
+
+  final override protected lazy val devMode = true
+  final override protected lazy val enableContractUpgrading = true
+
+  val languageVersion: LanguageVersion = LanguageVersion.v2_1
+  override val majorLanguageVersion: LanguageMajorVersion = languageVersion.major
+
+  override protected lazy val darFiles = List()
+
+  lazy val damlScriptDar = requiredResource("daml-script/daml3/daml3-script-2.1.dar")
+  lazy val upgradeTestLibDar: Path = rlocation(Paths.get("daml-script/test/upgrade-test-lib.dar"))
+
+  lazy val tempDir: Path = Files.createTempDirectory("upgrades-it")
+
+  val testFileDir: Path = rlocation(Paths.get("daml-script/test/daml/upgrades/"))
+  val testCases: Seq[TestCase] = getTestCases(testFileDir)
+
+  private def traverseSequential[A, B](elems: Seq[A])(f: A => Future[B]): Future[Seq[B]] =
+    elems.foldLeft(Future.successful(Seq.empty[B])) { case (comp, elem) =>
+      comp.flatMap { elems => f(elem).map(elems :+ _) }
+    }
+
+  // Maybe provide our own tracer that doesn't tag, it makes the logs very long
+  "Multi-participant Daml Script Upgrades" should {
+    testCases.foreach { testCase =>
+      testCase.name in {
+        for {
+          // Build dars
+          (testDarPath, deps) <- Future { buildTestCaseDar(testCase) }
```

Comment:
~~Can we replace this with an `(testDarPath, deps) = ...` binding? I guess we'd have to use `_ <- Future(())` to satisfy scala's syntax restrictions - up to you stylistically.~~

I see that `buildTestCaseDar` throws exceptions for failed assertions. Maybe leave a comment so that nobody refactors this?

Reply:
I moved the `Future` application into `buildTestCaseDar`, which cleans this up a bit. Note that we want to refactor the test setup to happen in a beforeall so this will likely change again.